In [44]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [45]:
df=pd.read_csv('subscription_prediction.csv')

In [46]:
df.head(100)

age          job   marital            education  default  housing  \
0    40       admin.   married             basic.6y       no       no   
1    56     services   married          high.school       no       no   
2    41  blue-collar   married              unknown  unknown       no   
3    57    housemaid  divorced             basic.4y       no      yes   
4    39   management    single             basic.9y  unknown       no   
..  ...          ...       ...                  ...      ...      ...   
95   34     services    single          high.school       no  unknown   
96   46       admin.  divorced    university.degree       no      yes   
97   30   technician   married  professional.course       no       no   
98   35     services    single             basic.4y       no       no   
99   30       admin.    single    university.degree       no       no   

       loan    contact month day_of_week  duration  campaign  pdays  previous  \
0        no  telephone   may         mon       151         1    999         0   
1       yes  telephone   may         mon       307         1    999         0   
2        no  telephone   may         mon       217         1    999         0   
3        no  telephone   may         mon       293         1    999         0   
4        no  telephone   may         mon       195         1    999         0   
..      ...        ...   ...         ...       ...       ...    ...       ...   
95  unknown  telephone   may         tue       147         2    999         0   
96      yes  telephone   may         tue       230         1    999         0   
97       no  telephone   may         tue       252         1    999         0   
98       no  telephone   may         tue       329         1    999         0   
99       no  telephone   may         tue       246         1    999         0   

       poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0   nonexistent           1.1          93.994          -36.4      4.857   
1   nonexistent           1.1          93.994          -36.4      4.857   
2   nonexistent           1.1          93.994          -36.4      4.857   
3   nonexistent           1.1          93.994          -36.4      4.857   
4   nonexistent           1.1          93.994          -36.4      4.857   
..          ...           ...             ...            ...        ...   
95  nonexistent           1.1          93.994          -36.4      4.857   
96  nonexistent           1.1          93.994          -36.4      4.857   
97  nonexistent           1.1          93.994          -36.4      4.857   
98  nonexistent           1.1          93.994          -36.4      4.857   
99  nonexistent           1.1          93.994          -36.4      4.857   

    nr.employed   y  
0        5191.0  no  
1        5191.0  no  
2        5191.0  no  
3        5191.0  no  
4        5191.0  no  
..          ...  ..  
95       5191.0  no  
96       5191.0  no  
97       5191.0  no  
98       5191.0  no  
99       5191.0  no  

[100 rows x 21 columns]

In [47]:
df.shape

(10122, 21)

In [48]:
df.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [49]:
pd.set_option('display.max_columns', None)


In [50]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

- age: (numeric)
- job: type of job (categorical: 'admin','blue collar','entrepreneur','housemaid','management','retired','self employed','services','student','technician','unemployed','unknown')
- marital: marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
- education: (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
- default: has credit in default? (categorical: 'no','yes','unknown')
- housing: has housing loan? (categorical: 'no','yes','unknown')
- loan: has personal loan? (categorical: 'no','yes','unknown')
- contact: contact communication type (categorical: 'cellular','telephone')
- month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
- day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
- duration: last contact duration, in seconds (numeric).
- campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
- pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
- previous: number of contacts performed before this campaign and for this client (numeric)
- poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
- emp.var.rate: employment variation rate - quarterly indicator (numeric)
- cons.price.idx: consumer price index - monthly indicator (numeric)
- cons.conf.idx: consumer confidence index - monthly indicator (numeric)
- euribor3m: euribor 3 month rate - daily indicator (numeric)
nr.employed: number of employees - quarterly indicator (numeric)

In [51]:
df['loan'].value_counts()

loan
no         8317
yes        1557
unknown     248
Name: count, dtype: int64

In [52]:
df['marital'].value_counts()

marital
married     5864
single      3152
divorced    1082
unknown       24
Name: count, dtype: int64

In [53]:
df['job'].unique()

array(['admin.', 'services', 'blue-collar', 'housemaid', 'management',
       'technician', 'entrepreneur', 'retired', 'student',
       'self-employed', 'unknown', 'unemployed'], dtype=object)

In [54]:
df['education'].unique()

array(['basic.6y', 'high.school', 'unknown', 'basic.4y', 'basic.9y',
       'university.degree', 'professional.course', 'illiterate'],
      dtype=object)

In [55]:
df['default'].unique()

array(['no', 'unknown'], dtype=object)

In [56]:
df['contact'].unique()

array(['telephone', 'cellular'], dtype=object)

In [57]:
df['housing'].unique()

array(['no', 'yes', 'unknown'], dtype=object)

In [58]:
# this is the company that is lunch the data

In [59]:
df['month'].unique()

array(['may', 'jun', 'jul', 'aug', 'oct', 'nov', 'dec', 'mar', 'apr',
       'sep'], dtype=object)

In [60]:
df['day_of_week'].unique()

array(['mon', 'tue', 'wed', 'thu', 'fri'], dtype=object)

In [61]:
date=df['duration']//60
date.value_counts()

duration
2     1543
1     1424
3     1278
4      967
5      709
0      574
6      572
7      416
8      367
9      325
10     294
11     272
12     204
13     180
14     163
15     132
16     107
17     104
18      83
20      62
19      54
22      52
21      43
24      32
25      24
23      23
26      20
28      14
27      11
30      10
32      10
33       8
34       6
31       6
41       4
40       4
44       3
29       3
36       2
60       2
35       2
38       2
51       2
37       1
53       1
43       1
46       1
69       1
39       1
54       1
48       1
63       1
Name: count, dtype: int64

In [62]:
df['campaign'].unique()

array([ 1,  2,  3,  4,  5,  7,  6,  9, 12, 23,  8, 11, 16, 20, 19, 39, 10,
       42, 15, 26, 28, 32, 13, 24, 17, 14, 22, 29, 30, 18, 41, 31, 35, 21,
       27])

In [63]:
df['pdays'].unique() # last the previous called 

array([999,   6,   4,   5,  10,   7,   1,  11,   9,   2,   0,   3,  12,
        13,  14,  15,  16,   8,  21,  18,  22,  25,  26,  17,  27,  19])

In [64]:
df['education'].value_counts()

education
university.degree      3201
high.school            2317
professional.course    1325
basic.9y               1316
basic.4y                989
basic.6y                498
unknown                 471
illiterate                5
Name: count, dtype: int64

In [65]:
# df['y'].apply(map({'Yes':1, 'No':0}))
df['y']=df['y'].apply(lambda x :1 if x=='yes' else 0)

In [66]:
df['y'].unique()

array([0, 1])

# divided the data in train test-->

In [67]:
# data set , to divide the algorith 

In [68]:
train_data=df.sample(frac=.85,random_state=42)

In [69]:
train_data.shape

(8604, 21)

In [70]:
# find the rations of the data
df['y'].value_counts(normalize=True)

y
0    0.541593
1    0.458407
Name: proportion, dtype: float64

In [71]:
train_data['y'].value_counts(normalize=True)

y
0    0.540679
1    0.459321
Name: proportion, dtype: float64

In [72]:
test_data = df.drop(train_data.index)

## here to find the diffrence the correct or not 

In [73]:
X_train = train_data.drop('y',axis=1)
y_train= train_data['y']
X_test=test_data.drop('y',axis=1)
y_test= test_data['y']

In [74]:
# KNN for one feature s
#  manhttan distance 
#  minlowski disance
# hamming distance

In [75]:
from sklearn.neighbors import KNeighborsClassifier


In [76]:
df['age']

0        40
1        56
2        41
3        57
4        39
         ..
10117    64
10118    37
10119    73
10120    44
10121    74
Name: age, Length: 10122, dtype: int64

In [102]:
def Knn(feature, single_row,k): # feature means have the columsn single
    diff=df[feature]-single_row[feature]
    smallest_k = diff.nsmallest(k)
    prediction=y_train[smallest_k.index].mode()[0]
    return prediction

In [103]:
Knn('age',X_train.iloc[101],5)

np.int64(1)

In [104]:
y_predict=X_test.apply(lambda x:Knn('age',x,4),axis=1)

In [105]:
(y_test==y_predict).sum()/len(y_test)*100

np.float64(45.3227931488801)